add namespace cmd arg

andyxning opened the pull request:
Fix #152

This PR add `namespaces` command line argument to kube-state-metrics. It's value is in the same format as `collectors`. If specified with delimited namespaces, kube-state-metrics will only collect resources in those namespaces.

/cc @brancz @matthiasr

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kube-state-metrics/201)
<!-- Reviewable:end -->

brancz commented:
There are loads of things going on here, let's first look at the problem again.

Issue #152 describes that the problem is that the caches are actually trying to request objects from all namespaces, however, this implementation still requests objects from all namespaces, but then filters those out that are not in the given namespace.

The solution is actually much more simple than what you proposed here. The caches already take the namespace they should run in as an argument, so all that has to be done is pass the namespace from a flag, and then pass it to the collector.

Aside from that I appreciate the attempt to refactor some things, but I'd prefer to separate refactoring from feature development (at least by commits). In terms of this feature though I'd prefer to keep it as much on track as possible and avoid the refactoring here as all of this is very core to kube-state-metrics working reliably.

andyxning commented:
Yep. @brancz you are right. I have implemented it wrongly. :( 

Will correct it later.

andyxning commented:
@brancz I have updated the implementation. PTAL.

BTW, in the newest implementation, i have updated all namespace argument list and watch except `node` which is not binding to any namespace.

andyxning commented:
Maybe this pr will also be accept after 1.0 GA. @brancz

brancz commented:
lgtm now :+1: nice! Yes we'll hold off until the 1.0. Nice job though.

zouyee commented:
sounds good, something like that [`Namespace to watch for scrapy. Default is to watch all namespaces` ](https://github.com/kubernetes/ingress/tree/master/controllers/nginx#command-line-arguments)

andyxning commented:
@brancz I have rebased the code. PTAL.

brancz commented:
Turns out the wonderful interface for the register funcitons is not as good as I thought :slightly_smiling_face: that's totally not your fault but if it becomes confusing we might want to think about functional arguments or something like that to make this more clean.

But these changes lgtm here :+1:
